### PR TITLE
Add Bengali bn

### PR DESCRIPTION
**Pull request by arijitx**
Basic Rules for Bengali, updated segmenter with nltk PunktSentenceTokenizer as punkt doesnt have a pretrained model in nltk for bengali so used the sentence end as '।'

How many sentences did you get at the end?
~ 129000

How did you create the blacklist file?
Followed instruction in readme, chose words with frequency 50 or less

Evaluate random 500 sentences ?
[500 random sentences](https://docs.google.com/spreadsheets/d/1BJFLOq3L1gDYnHfMVSfUH_r43HS2HspDfrT0Gv3DREw/edit?usp=sharing)
I do not know any linguist any Bengali speaker please feel free to comment on the error ratio. My analysis shows very less number of errors in the extracted sentences.

### CODE REVIEW

**Comment by MichaelKohler** [comment]
Thanks for working on this. I will have a few comments, will have a closer look later tonight or tomorrow.

Regarding linguists, would be great, but it's fine if it's "just" native speakers. For the amount, it should be more than 100 sentences though. I would suggest 500. Once I have added my comments, there is some work to be done, so I would suggest to wait with the review until then.

**Comment by arijitx** [comment]
Sure @MichaelKohler waiting for your comments

File: src/rules/bn.toml [hunk 1]
```diff
@@ -0,0 +1,39 @@
+
+# Replacements which are run before sentences are split
+replacements = []
+
+min_trimmed_length = 3
+min_word_count = 3
```
Comment (MichaelKohler) [inline src/rules/bn.toml:6]: Unlikely, but just to make sure: there are no useful sentences with 2 or less words in Wikipedia. Right?

File: src/segmenter.rs [hunk 2]
```diff
@@ -32,6 +33,7 @@ pub fn split_sentences_with_python_en(text: &str) -> Vec<String> {
     ctx.get("split_sentences")
 }
 
+
```
Comment (MichaelKohler) [inline src/segmenter.rs:36]: minor: please remove this new linebreak

File: src/segmenter.rs [hunk 3]
```diff
@@ -70,3 +101,5 @@ mod test {
         assert_eq!(split_sentences_with_python(language, text).len(), 2);
     }
 }
+
+
```
Comment (MichaelKohler) [inline src/segmenter.rs:105]: minor: please remove this new linebreak

**Review by MichaelKohler — CHANGES_REQUESTED** [review]
Thanks again! I've added a few remarks and questions :)

File: src/rules/bn.toml (hunk 1 above)
Comment (arijitx) [reply src/rules/bn.toml:6]: It's rare, tough there can be meaningful phrases as well as entities which can make sense. I will try out with 2 min words and check if that add reasonable words in the data.

File: src/rules/bn.toml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+]
+
+even_symbols = ['"']
+
+matching_symbols = [
+  ["„", "“"],
+  ["(", ")"]
+]
+
+abbreviation_patterns = [
```
[the remark this answers is not in the document]
Comment (arijitx) [reply src/rules/bn.toml:35]: Updated, I thought as most abbr. in Bengali phonetic in nature, I will update with some rules following en rules.

File: src/rules/bn.toml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+max_word_count = 20
+min_characters = 5
+may_end_with_colon = false
+quote_start_with_letter = true
+needs_punctuation_end = false
+needs_letter_start = true
+needs_uppercase_start = false
+broken_whitespace = ["  ", " ,", " .", " ?", " !", " ;"]
+
+disallowed_symbols = [
```
[the remark this answers is not in the document]
Comment (arijitx) [reply src/rules/bn.toml:16]: Thanks for the suggestion, yes, it sounds a better way to add with allowed_symbols_regex.

File: src/rules/bn.toml [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+may_end_with_colon = false
+quote_start_with_letter = true
+needs_punctuation_end = false
+needs_letter_start = true
+needs_uppercase_start = false
+broken_whitespace = ["  ", " ,", " .", " ?", " !", " ;"]
+
+allowed_symbols_regex = "[॰ঁংঃঅআইঈউঊঋঌএঐওঔকখগঘঙচছজঝঞটঠডঢণতথদধনপফবভমযরলশষসহ়ঽািীুূৃৄেৈোৌ্ৎৗয়ৠৡ৤০১২৩৪৫৬৭৮৯ৰৱ৳৹৺ਂ!\"',-\\./:;\\?`।„“‚‘’ ]"
+
+disallowed_symbols = [
```
Comment (MichaelKohler) [inline src/rules/bn.toml:28]: This can now be completely removed

File: src/rules/bn.toml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+even_symbols = ['"']
+
+matching_symbols = [
+  ["„", "“"],
+  ["(", ")"]
+]
+
+abbreviation_patterns = [
+  "([\u0995-\u09B9\u09CE\u09DC-\u09DF\u0985-\u0994\u09BE-\u09CC\u09D7\u09BC]+\\.+)+[\u0995-\u09B9\u09CE\u09DC-\u09DF\u0985-\u0994\u09BE-\u09CC\u09D7\u09BC]*"
```
Comment (MichaelKohler) [inline src/rules/bn.toml:48]: Could you add a comment to better explain what it does? I think that would be useful for whoever looks at it in the future. Example here: https://github.com/common-voice/cv-sentence-extractor/blob/main/src/rules/en.toml#L36

**Review by MichaelKohler — CHANGES_REQUESTED** [review]
Two more minor comments, the rest looks good to me.

**Comment by arijitx** [comment]
Hi @MichaelKohler I have applied your suggest changes, please feel free to review, I have also created another set of 500 sentences with the updated rules. Do let me know if am good to start the review.

File: src/rules/bn.toml (hunk 6 above)
Comment (arijitx) [reply src/rules/bn.toml:28]: Is it okay to keep digits in the allowed_symbols_regex ?

File: src/rules/bn.toml (hunk 7 above)
Comment (arijitx) [reply src/rules/bn.toml:40]: Thanks for suggestion, have added some comments.

File: src/rules/bn.toml (hunk 6 above)
Comment (MichaelKohler) [reply src/rules/bn.toml]: Oh, good question! We don't allow numbers in general, as it's not necessarily clear how to pronounce them. 2022 could for example be "two-zero-two-two", "twenty-twenty-two" or "two thousand and twenty two". Is that the same for Bengali? If so, let's not allow numbers.

Comment (arijitx) [reply src/rules/bn.toml]: Got it my bad. I will update the rules to exclude numbers

**Comment by arijitx** [comment]
Thanks for your quick reviews @MichaelKohler do let me know if am good to start with the error rate review process ?

File: src/segmenter.rs [hunk 8]
```diff
@@ -61,6 +84,13 @@ mod test {
         assert_eq!(split_sentences_with_python(language, text).len(), 2);
     }
 
+    fn test_segmenter_bn() {
```
Comment (MichaelKohler) [inline src/segmenter.rs:87]:
```suggestion
    #[test]
    fn test_segmenter_bn() {
```

**Review by MichaelKohler — CHANGES_REQUESTED** [review]
Thanks for the changes! This looks good apart from one more comment I missed before. However this does not block any review. At the bottom of https://github.com/common-voice/cv-sentence-extractor/actions/runs/1674625643 you find a "extraction" artifact. That zip file contains a `sample.txt` which you can use for the review. I would suggest to shuffle this file randomly, to make sure that extractions are not all from the same article (`shuf sample.txt > shuffled.txt`).

**Comment by arijitx** [comment]
Thanks @MichaelKohler , I have downloaded the sample and shuffled it randomly and taken 500 for further review.

**Review by MichaelKohler — APPROVED** [review]
Looks good to me! Approving the code changes and looking forward to the result of the review!

**Comment by arijitx** [comment]
Hi @MichaelKohler,

I could arrange two reviews, here are the results still WIP for the 3rd review

### Review 1:
**Correct rate: 95%**
https://docs.google.com/spreadsheets/d/1BJFLOq3L1gDYnHfMVSfUH_r43HS2HspDfrT0Gv3DREw/edit?usp=sharing

### Review 2:
**Correct rate 96%**
https://docs.google.com/spreadsheets/d/1XPjdZzyzqrXSCwTJvAtD6sK3L-00i5CERz-QBTk0MfQ/edit?usp=sharing

**Comment by MichaelKohler** [comment]
Now that you have the reviews, did ypu notice a pattern in the wrong sentences? Anything that easily could be filtered out? If so, would be great to add that and do another review. If not, also ok :)

**Comment by arijitx** [comment]
Hi @MichaelKohler , From my analysis most of the error is like incomplete/partial sentence or phrases, there isn't a very high number of them, either its written that way and probably from the context of article it made sense. I don't see a clear way to filter such sentences.

**Review by MichaelKohler — APPROVED** [review]
Looking all good then. Thanks for working on this! I will start the extract and CC you on the Common Voice PR I'm gonna create.

**Pull request merged**